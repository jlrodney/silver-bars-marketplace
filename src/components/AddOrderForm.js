import React from 'react';
import uuidv1 from 'uuid/v1';
import FormErrors from './FormErrors';

export default class AddOrderForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      orderId: uuidv1(),
      userId: '',
      quantity: '',
      type: 'BUY',
      price: '',
      noErrors: false,
      formErrors: {
        parameters: '',
        positiveNumber: '',
        stringCheck: '',
        typeCheck: ''
      }
    };
  }

  onUserIdChange = (e) => {
    const userId = e.target.value;
    this.setState(() => ({ userId }));
  };
  onQuantityChange = (e) => {
    let quantity = String(e.target.value);
    if ( !quantity || quantity.match(/^\d{1,}(\.\d{0,8})?$/) ) {
      this.setState(() => ({ quantity }));
    }
  };

  onPriceChange = (e) => {
    const price = e.target.value;
    if ( !price || price.match(/^\d{1,}(\.\d{0,2})?$/) ) {
      this.setState(() => ({ price }));
    }
  };
  onTypeChange = (e) => {
    const type = e.target.value;
    this.setState(() => ({ type }));
  };

  validateInputs = async () => {
      await this.typeCheck();
      await this.parametersCheck();
      await this.positiveNumber();
      await this.stringCheck();
      this.errorsChecker();
  }

  parametersCheck = () => {
    if (!this.state.userId || !this.state.price || !this.state.quantity || !this.state.type) {
      this.setState((prevState) => ({ formErrors: {
        ...prevState.formErrors,
        parameters: 'Please fill in all fields of the form. ' }}));
    } else {
      this.setState((prevState) => ({ formErrors: {
         ...prevState.formErrors,
         parameters: ''
       }}))
    }
  }


  positiveNumber = () => {
    if (parseFloat(this.state.price, 10) <= 0 || parseFloat(this.state.quantity, 10) <= 0) {
      this.setState((prevState) => ({ formErrors: {
        ...prevState.formErrors,
        positiveNumber: 'Please ensure both quantity and price are positive numbers. ' }}))
    } else {
      this.setState((prevState) => ({ formErrors: {
         ...prevState.formErrors,
         positiveNumber: ''
       }}))
    }
  }

  stringCheck = () => {
    if ( typeof(this.state.userId) !== "string" || typeof(this.state.type) !== "string") {
      this.setState((prevState) => ({ formErrors: {
        ...prevState.formErrors,
        stringCheck: 'Please ensure both userid and order type are strings. ' }}))
    } else {
      this.setState((prevState) => ({ formErrors: {
         ...prevState.formErrors,
         stringCheck: ''
       }}))
    }
  }

  typeCheck = () => {
    if (this.state.type !== "BUY" && this.state.type !== "SELL") {
      this.setState((prevState) => ({ formErrors: {
        ...prevState.formErrors,
        typeCheck: 'Please ensure type is either "BUY" or "SELL". ' }}))
    } else {
      this.setState((prevState) => ({ formErrors: {
         ...prevState.formErrors,
         typeCheck: ''
       }}))
    }
  }

  errorsChecker =  () => {
     this.noErrors(Object.keys(this.state.formErrors).every((key) => this.state.formErrors[key] === '' ));
  }

  noErrors = (noErrors) => {
     this.setState(() => ( { noErrors }))
  }

  onSubmit =  async (e) => {
    e.preventDefault();
    const orderId = uuidv1();
    this.setState(() => ({ orderId }));
    await this.validateInputs();
    if ( this.state.noErrors === true ) {
      this.props.onSubmit({
        orderId: this.state.orderId,
        userId: this.state.userId,
        price: parseFloat(this.state.price, 10).toFixed(2),
        type: this.state.type,
        quantity: parseFloat(this.state.quantity, 10)
      });
    }
  };
  render() {
    return (
      <div>
        <FormErrors formErrors={this.state.formErrors}/>
        <form onSubmit={this.onSubmit}>
        <label htmlFor="userId">User id: </label>
          <input
            name="userId"
            type="text"
            placeholder="UserId"
            autoFocus
            value={this.state.userId}
            onChange={this.onUserIdChange}
          />
          <label htmlFor="price"> Price (£/kg): </label>
          <input
            name="price"
            type="text"
            placeholder="Price £/kg"
            value={this.state.price}
            onChange={this.onPriceChange}
          />
          <label htmlFor="quantity"> Quantity (kg): </label>
          <input
          name="quantity"
          type="text"
          placeholder="Quantity (kg)"
          value={this.state.quantity}
          onChange={this.onQuantityChange}
          >
          </input>
          <select name="type" onChange={this.onTypeChange} value={this.state.type}>
            <option value="BUY">BUY</option>
            <option value="SELL">SELL</option>
          </select>
          <button className="button" >Place Order</button>
        </form>
      </div>
    )
  }
}
