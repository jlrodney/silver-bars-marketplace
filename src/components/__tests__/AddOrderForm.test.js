import React from 'react';
import { shallow } from 'enzyme';
import AddOrderForm from '../AddOrderForm';

const order = {
  orderId: 'xyz987',
  userId: '1',
  quantity: 22,
  price: 230,
  type: 'BUY',
  error: ''
}
describe('Add order form component', () => {
  it('should render form', () => {
    const wrapper = shallow(<AddOrderForm />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render form correctly with order', () => {
    const wrapper = shallow(<AddOrderForm order={order} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render place order button', () => {
    const wrapper = shallow(<AddOrderForm />);
    expect(wrapper.find('button').text()).toEqual("Place Order");
  });

  it('should render error for invalid form submission', () => {
    const wrapper = shallow(<AddOrderForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { }
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('should set description on input change', () => {
    const value = 'user2';
    const wrapper = shallow(<AddOrderForm />);
    wrapper.find('input').at(0).simulate('change', {
      target: { value }
    });
    expect(wrapper.state('userId')).toBe(value);
  });

  it('should set quantity on quantity change', () => {
    const value = "3";
    const wrapper = shallow(<AddOrderForm />);
    wrapper.find('input').at(2).simulate('change', {
      target: { value }
    });
    expect(wrapper.state('quantity')).toBe(value);
  });

  it('should set price if valid input', () => {
    const value = '23.50';
    const wrapper = shallow(<AddOrderForm />);
    wrapper.find('input').at(1).simulate('change', {
      target: { value }
    });
    expect(wrapper.state('price')).toBe(value);
  });

  it('should not set price if invalid input', () => {
    const value = '12.122';
    const wrapper = shallow(<AddOrderForm />);
    wrapper.find('input').at(1).simulate('change', {
      target: { value }
    });
    expect(wrapper.state('price')).toBe('');
  });

  it('should call onSubmit prop for invalid form submission', () => {
    const wrapper = shallow(<AddOrderForm />);
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { }
    });
    expect(wrapper.state('error')).toBe("Please fill in all fields of the form.");
  });
  it('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<AddOrderForm onSubmit={onSubmitSpy} />);
    let value = 'user1';
    wrapper.setState({ orderId: '141'})
    wrapper.find('input').at(0).simulate('change', {
      target: { value }
    });
    value = '33';
    wrapper.find('input').at(1).simulate('change', {
      target: { value }
    });
    value = 320;
    wrapper.find('input').at(2).simulate('change', {
      target: { value }
    });
    value = "SELL";
    wrapper.find('select').simulate('change', {
      target: { value }
    });
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { }
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
      orderId: '141',
      userId: 'user1',
      quantity: 320,
      type: 'SELL',
      price: 3300,
    });
  });
});
