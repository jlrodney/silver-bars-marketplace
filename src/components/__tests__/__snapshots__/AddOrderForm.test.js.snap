// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add order form component should render form 1`] = `
<div>
  <FormErrors
    formErrors={
      Object {
        "parameters": "",
        "positiveNumber": "",
        "stringCheck": "",
        "typeCheck": "",
      }
    }
  />
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="userId"
    >
      User id: 
    </label>
    <input
      autoFocus={true}
      name="userId"
      onChange={[Function]}
      placeholder="UserId"
      type="text"
      value=""
    />
    <label
      htmlFor="price"
    >
       Price (£/kg): 
    </label>
    <input
      name="price"
      onChange={[Function]}
      placeholder="Price £/kg"
      type="text"
      value=""
    />
    <label
      htmlFor="quantity"
    >
       Quantity (kg): 
    </label>
    <input
      name="quantity"
      onChange={[Function]}
      placeholder="Quantity (kg)"
      type="text"
      value=""
    />
    <select
      name="type"
      onChange={[Function]}
      value="BUY"
    >
      <option
        value="BUY"
      >
        BUY
      </option>
      <option
        value="SELL"
      >
        SELL
      </option>
    </select>
    <button
      className="button"
    >
      Place Order
    </button>
  </form>
</div>
`;

exports[`Add order form component should render form correctly with order 1`] = `
<div>
  <FormErrors
    formErrors={
      Object {
        "parameters": "",
        "positiveNumber": "",
        "stringCheck": "",
        "typeCheck": "",
      }
    }
  />
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="userId"
    >
      User id: 
    </label>
    <input
      autoFocus={true}
      name="userId"
      onChange={[Function]}
      placeholder="UserId"
      type="text"
      value=""
    />
    <label
      htmlFor="price"
    >
       Price (£/kg): 
    </label>
    <input
      name="price"
      onChange={[Function]}
      placeholder="Price £/kg"
      type="text"
      value=""
    />
    <label
      htmlFor="quantity"
    >
       Quantity (kg): 
    </label>
    <input
      name="quantity"
      onChange={[Function]}
      placeholder="Quantity (kg)"
      type="text"
      value=""
    />
    <select
      name="type"
      onChange={[Function]}
      value="BUY"
    >
      <option
        value="BUY"
      >
        BUY
      </option>
      <option
        value="SELL"
      >
        SELL
      </option>
    </select>
    <button
      className="button"
    >
      Place Order
    </button>
  </form>
</div>
`;
